File: practice/algorithms/search/cut_the_tree.cpp
#include <unordered_map>
#include <iostream>
#include <vector>
using namespace std;


long calcSums(int start, unordered_map<int, vector<int>> &graph, int* values, 
		int n, unordered_map<int, int>& parent, long* best) {

	int neib;
	long val = values[start];
	vector<int> neibs = graph.at(start);
	for (int i = 0; i < neibs.size(); i++) {
		neib = neibs[i];
		if (parent.at(start) != neibs[i]) {
			parent.insert({neib, start});
			val += calcSums(neib, graph, values, n, parent, best);
		}
	}
	best[start] = val;
	return val;
}
		

void solution(unordered_map<int, vector<int>>& graph, int* values, int n) {
	long best[n];
	unordered_map<int, int> parent;
	parent.insert({0,-1});
	calcSums(0, graph, values, n, parent, best);

	long m;
	m = abs(best[0] - best[1] - best[1]);
	for (int i = 2; i < n; i++) {
		m = min(m, abs(best[0] - best[i] - best[i]));
	}
	cout << m;
}


int main() {
	
	int n, val, u, v;
	cin >> n;
	int values[n];
	unordered_map<int, vector<int>> graph;
	for (int i = 0; i < n; i++) {
		cin >> val;
		values[i] = val;
		vector<int> empty;
		pair<int, vector<int>> p (i, empty);
		graph.insert(p);
	}
	for (int i = 0; i < n-1; i++) {
		cin >> u >> v;
		u -= 1; v -= 1;
		graph.at(u).push_back(v);
		graph.at(v).push_back(u);
	}

	solution(graph, values, n);
	
	return 0;
}

File: practice/algorithms/search/cut_the_tree.py
#https://www.hackerrank.com/challenges/cut-the-tree/problem

import sys
sys.setrecursionlimit(10000000)

def calcSums(start, graph, values, n, best, parent):
    val = values[start]
    for neib in graph[start]:
        if parent[start] != neib:
            parent[neib] = start
            val += calcSums(neib, graph, values, n, best, parent)
    best[start] = val
    return val

def solution(graph, values, n):
    best = [0 for i in range(n)]
    parent = {}
    parent[0] = -1
    calcSums(0, graph, values, n, best, parent)
    
    minn = None
    for i in range(1, n):
        if minn == None:
            minn = abs((best[0] - best[i]) - best[i])
        else:
            minn = min(minn, abs((best[0] - best[i]) - best[i]))
    return minn

if __name__ == "__main__":
    n = int(input())
    values = [int(_) for _ in input().strip().split()]
    graph = {i : [] for i in range(n)}
    for i in range(n-1):
        u, v = [int(_) for _ in input().strip().split()]
        graph[u-1].append(v-1)
        graph[v-1].append(u-1)
    ans = solution(graph, values, n)
    print(ans)
